account/src/main/java/com/caldevsupplychain/account/controller/AccountControllerImpl.java

```diff
@@ -84,24 +86,161 @@
 			return new ResponseEntity<>(new ApiErrorsWS(ErrorCode.ACCOUNT_EXIST.name(), "Account already registered."), HttpStatus.CONFLICT);
 		}
 
-		UserBean userBean = userMapper.userWSToBean(userWS);
+		UserBean userBean = userMapper.toBean(userWS);
 
 		UserBean user = accountService.createUser(userBean);
 
 		try {
-			emailService.sendVerificationTokenEmail(user.getEmailAddress(), user.getToken(), EmailType.REGISTRATION.name());
+			emailService.sendVerificationTokenEmail(user.getEmailAddress(), user.getToken(), EmailType.ACTIVATING.name());
 		} catch (MessagingException e) {
 			return new ResponseEntity<>(new ApiErrorsWS(ErrorCode.EMAIL_MESSAGING_EXCEPTION.name(), e.getMessage()), HttpStatus.INTERNAL_SERVER_ERROR);
 		}
+		return new ResponseEntity<>(userMapper.toWS(user), HttpStatus.CREATED);
+	}
+
+	@GetMapping("/activate/{token}")
+	public ResponseEntity<?> activateAccount(@PathVariable("token") String token) {
+		Optional<UserBean> user = accountService.findByToken(token);
+
+		if (!user.isPresent()) {
+			return new ResponseEntity<Object>(new ApiErrorsWS(ErrorCode.INVALID_TOKEN.name(), "Invalid Token."), HttpStatus.BAD_REQUEST);
+		}
+		accountService.activateUser(user.get().getId());
 
-		return new ResponseEntity<>(userMapper.userBeanToWS(user), HttpStatus.CREATED);
+		// TODO: need to tell FE to redirect to /login page after activate signup token
+		return new ResponseEntity<>(userMapper.toWS(user.get()), HttpStatus.OK);
 	}
 
+	@PostMapping("/login")
+	public ResponseEntity<?> login(@Validated @RequestBody UserWS userWS) {
+
+		Optional<UserBean> user = accountService.findByEmailAddress(userWS.getEmailAddress());
+
+		if (!user.isPresent()) {
+			return new ResponseEntity<>(new ApiErrorsWS(ErrorCode.LOGIN_INVALID.name(), "Invalid Login"), HttpStatus.UNAUTHORIZED);
+		}
+
+		Subject subject = SecurityUtils.getSubject();
+
+		// Apache Shiro authentication check for 1st login
+		if (!subject.isAuthenticated()) {
+			UsernamePasswordToken token = new UsernamePasswordToken(userWS.getEmailAddress(), userWS.getPassword());
+
+			try {
+				subject.login(token);
+			} catch (AuthenticationException e) {
+				log.error("Error in login. User fail to login. user={}", user.toString());
+				return new ResponseEntity<>(new ApiErrorsWS(ErrorCode.LOGIN_INVALID.name(), e.getMessage()), HttpStatus.UNAUTHORIZED);
+			}
+		}
+
+		log.info("Success in user login. user={}", user.toString());
+
+		UserBean userBean = user.get();
+		JwtBean jwtBean = jwtMapper.toBean(userBean);
+		JWTAuthenticationToken jwtObj = jwtService.createJwtToken(jwtBean);
+
+		log.info("Success create jwt token. jwtToken={}", jwtObj.getToken());
+
+		return new ResponseEntity<>(userMapper.toWS(userBean), jwtService.createJwtHeader(jwtObj.getToken()), HttpStatus.OK);
```

Comment:
no need to return UserWS here.

Reply:
fix

Reply:
Should I still return responseBody about user information when they successfully login? This will be cache as a "state" in FE part and these data can be used to display when they login and redirect to a dashboard or something. Let me know what you think about this and I will refactor the code.